woocommerce/woocommerce-android
# Issue/3726 fetch order when payment flow starts

Parent issue #3726 

This PR makes sure we re-fetch order from the server before we try to collect a payment. This is necessary as unless we have up-to-date version of the order, we can't be sure the order hasn't been paid for already. 

I had to do more changes than I anticipated, I recommend reviewing this PR by commits so you can follow my train of thoughts. 

Merge instructions:
1. Make sure https://github.com/woocommerce/woocommerce-android/pull/4219 is merged
2. Remove "do not merge" label
3. Merge this PR

52a2d68 - fetches order from the server instead of loading it from the DB when the flow starts. LoadFromDb() was returning WCOrderModel but fetchOrder returns Order (app model). All the changes in this commit are related to this.
8b14774 - shows error state when fetching order fails. I introduced new `PaymentFlowError` since the previously used error type (CardPaymentStatusErrorType) lives in the CardReader module and fetching order is not related to that module. 
99bcc3b - we fetch order twice during the payment flow -> once when the flow starts, this is introduced in this PR and once when the payment succeeds to make sure the order detail displays "paid" state of the order. This commit renames the fields/methods so "fetchOrder" is used for the first case and "reFetchOrder" is used for the second case.
c992ce3 - added test for the "fetching order failed" error state
d457bd5 - removes fields which are not used anymore
964ffd9 - removes redundant let
c719324 - automatically finishes the flow and shows a snackbar when the order is already in paid state (solution updated in 3508bb2)
4a74c81 - removes try/catch as we removed IllegalStateExceptions in 8b14774 and 52a2d68
cb2de08 - reverts a change I didn't do on purpose (thank you unit tests ;) )
01fe5be - removes unused app logger
db7dc70 - fetchOrder was returning a cached value when the request to the backend failed. This it not the desired behavior for this flow as we want to make sure not to create duplicate payments for paid orders.
717c798 - I noticed that when you click on "retry" the request fails so quickly that it's not even noticable -> added a synthetic delay
cc724e6 - Fixes all usages of fetchOrder in tests as they don't expect the newly introduced parameter
7f4954f, 09ff9b4 - these commits were cherry-picked into [a standalone PR ](https://github.com/woocommerce/woocommerce-android/pull/4219)to facilitate review of this PR
3508bb2 - starts using isCollectable check in the payment flow


To test
1. Connect to a reader (can be simulated)
2. Open detail of an unpaid order in US dollars
3. Turn on airplane mode
4. Tap on "Collect Payment"
5. Notice "Error fetching order" state shown
6. Press back to dismiss the error
2. Go to the wp-admin and mark the order as paid
3. Turn off the airplane mode
3. Tap on "Collect payment" in the app
4. Notice the payment fails and "Order already paid" snackbar is shown. Also notice the "Collect payment" button is hidden now.
-------------------------------------


Update release notes:
